### PR TITLE
Support for Custom Header Ids - replaces #553

### PR DESCRIPTION
**Pull request by robbear**
Replaces #553

Sometimes it's helpful to be able to specify a slug for a header in the Markdown source, e.g., to shorten it, to avoid collisions in a way that is robust to re-ordering, or to avoid having Unicode characters in URLs:

### This is a really long title for a long section [#long-section]

This PR adds `mdx-plugins/remark-header-ids` to support the above syntax and adds a `customId` field to the `Heading` type. It ensures that the table of contents items reflect links to custom ids. Examples are provided in the `custom-header-ids.*.mdx` files.

### CODE REVIEW

File: examples/swr-site/pages/docs/custom-header-ids.es-ES.mdx [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+## Foo [#bar]
+```
+
+## Esto Tiene una Identificación Personalizada [#ejemplo-id-personalizado]
+
+Puede ver la identificación del encabezado personalizado en el encabezado de arriba. Tenga en cuenta tanto al pasar el cursor sobre el encabezado como en la tabla de contenido, el href `ejemplo-id-personalizado` está en uso.
+
+## Esto No Tiene
+
+Este encabezado no tiene una identificación personalizada. El id se genera a partir del texto del encabezado.
```
Comment (B2o5T) [inline examples/swr-site/pages/docs/custom-header-ids.es-ES.mdx:15]: we don't need translation for this change, it's better to add some test for this change

File: packages/nextra-theme-docs/src/components/toc.tsx [hunk 2]
```diff
@@ -70,10 +70,10 @@ export function TOC({ headings, filePath }: TOCProps): ReactElement {
             {renderComponent(config.toc.title)}
           </p>
           <ul>
-            {items.map(({ id, value, depth }) => (
-              <li className="nx-my-2 nx-scroll-my-6 nx-scroll-py-6" key={id}>
+            {items.map(({ customId, id, value, depth }) => (
```
[the remark this answers is not in the document]
Comment (robbear) [reply packages/nextra-theme-docs/src/components/toc.tsx:73]: I didn't know if you wanted to preserve the original `id` value, hence the separate `customId`. But, yes, we can do what you propose in https://github.com/shuding/nextra/pull/1645#discussion_r1138846376

File: examples/swr-site/pages/docs/custom-header-ids.es-ES.mdx (hunk 1 above)
Comment (robbear) [reply examples/swr-site/pages/docs/custom-header-ids.es-ES.mdx:15]: 1. Do you want the non-English pages removed?
2. Could you give me more details on the sort of test you'd like? Is it an enhancement to an existing test, or something new and stand-alone? I'm unfamiliar with your test framework but I can give it a go with some guidance.

Comment (B2o5T) [reply examples/swr-site/pages/docs/custom-header-ids.es-ES.mdx:15]: you can remove non-english pages and about tests take a look at this file https://github.com/shuding/nextra/blob/main/packages/nextra/__test__/compile.test.ts

Comment (robbear) [reply examples/swr-site/pages/docs/custom-header-ids.es-ES.mdx]: I've added tests and updated snapshot files. Again, I'm unfamiliar with the process so let me know if I need to take a different approach.

Comment (robbear) [reply examples/swr-site/pages/docs/custom-header-ids.es-ES.mdx]: Looking into the linter errors...

Comment (robbear) [reply examples/swr-site/pages/docs/custom-header-ids.es-ES.mdx]: Done

**Review by B2o5T — APPROVED** [review]
thanks for working on it

I prettified little bit and added more test cases

I also searched existed solution

https://github.com/imcuttle/remark-heading-id similar, works with `... {#my-heading}` syntax, doesn't support mdx2

https://github.com/Eyas/md-heading-id/tree/main/packages/remark-custom-heading-id similar `{#my-heading}` syntax, seems like support mdx2? looked in package.json dependencies

https://github.com/reactjs/react.dev/pull/4105 another solution with `{/*some-heading*/}` syntax

I would still prefer to stay with your implementation

P.S. I would also be appreciated it if you mentioned this feature in `nextra.site` (should be easy since you mostly described it in SWR example website)

**Comment by robbear** [comment]
> P.S. I would also be appreciated it if you mentioned this feature in nextra.site (should be easy since you mostly described it in SWR example website)

https://github.com/shuding/nextra/pull/1706

**Pull request merged**